Feature: Use different datadirs for different signets

BrandonOdiwuor opened the pull request:
Fixes https://github.com/bitcoin/bitcoin/issues/27494

When the `-signetchallenge` argument is provided, the hash-160 of the challenge is appended to the data directory name. This ensures that each signet has its own distinct data directory, following the naming convention `signet_XXXXXXX`.

Note:
The behaviour of the default signet remains unchanged

### Update:

When the `-signetchallenge` argument is provided, `the first 8 characters of the hash-160 of the challenge` are appended to the data directory name. This ensures that each signet has its own distinct data directory, following the naming convention `signet_XXXXXXX`.

BrandonOdiwuor committed:
Use distinct data dirs for each signet

If the -signetchallenge argument is provided,
the hash-160 (first 8 characters) of the challenge is appended to the datadir name,
resulting in unique data directories for each signet, i.e., signet_XXXXXXX.

1440000bytes commented:
Concept ACK

laanwj commented:
Concept ACK

tdb3 commented on src/common/args.cpp:
```diff
@@ -277,6 +278,18 @@ fs::path ArgsManager::GetPathArg(std::string arg, const fs::path& default_value)
+    if (!signet_challenge_str.empty()) {
+        std::vector<uint8_t> signet_challenge = ParseHex(signet_challenge_str);
+        const auto hash_160 = Hash160(signet_challenge);
+        return fs::PathFromString(base_data_dir + "-" + HexStr(hash_160));
```
The approach to use the 8-character message start (i.e. first 4 bytes of the SHA256 hash of the challenge) seems better to me as well.  More concise directory lengths with a reasonably low chance of collision between signets.

nit:  Recommend avoiding using `-` (dash) and instead using something like `_` (underscore) when building signet paths to prevent unexpected issues when including custom signet directory paths in command line arguments.

BrandonOdiwuor replied:
I have updated the PR to use the `first 8 bytes of the hash160` and replaced the `-` with `_` as suggested

tdb3 reviewed (commented):
Concept ACK.
Great feature to help prevent conflicting datadir conflicts when switching between signets.
Inline comments added.

Received failures for functional tests feature_signet.py and tool_signet_miner.py.
```
dev@bdev01:~/bitcoin$ test/functional/test_runner.py -u --cachedir=/mnt/tmp/cache --tmpdir=/mnt/tmp test/functional/feature_signet.py test/functional/tool_signet_miner.py 
Temporary test directory at /mnt/tmp/test_runner_₿_🏃_20240413_214701
Remaining jobs: [feature_signet.py, tool_signet_miner.py --legacy-wallet, tool_signet_miner.py --descriptors]
1/3 - feature_signet.py failed, Duration: 60 s                                                                          

stdout:
2024-04-14T01:47:02.221000Z TestFramework (INFO): PRNG seed is: 2979480965586364484
2024-04-14T01:47:02.222000Z TestFramework (INFO): Initializing test directory /mnt/tmp/test_runner_₿_🏃_20240413_214701/feature_signet_2
2024-04-14T01:48:02.321000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 565, in start_nodes
    node.wait_for_rpc_connection()
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 314, in wait_for_rpc_connection
    self._raise_assertion_error("Unable to connect to bitcoind after {}s".format(self.rpc_timeout))
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 188, in _raise_assertion_error
    raise AssertionError(self._node_msg(msg))
AssertionError: [node 0] Unable to connect to bitcoind after 60s

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 131, in main
    self.setup()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 302, in setup
    self.setup_network()
  File "/home/dev/bitcoin/test/functional/feature_signet.py", line 43, in setup_network
    self.setup_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 418, in setup_nodes
    self.start_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 568, in start_nodes
    self.stop_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 583, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 375, in stop_node
    self.stop(wait=wait)
    ^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 205, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
                                  ^^^^^^^^^^^^^^^^^^^^
AssertionError: [node 0] Error: no RPC connection
2024-04-14T01:48:02.376000Z TestFramework (INFO): Stopping nodes
[node 5] Cleaning up leftover process
[node 4] Cleaning up leftover process
[node 3] Cleaning up leftover process
[node 2] Cleaning up leftover process
[node 1] Cleaning up leftover process
[node 0] Cleaning up leftover process


stderr:
Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/feature_signet.py", line 85, in <module>
    SignetBasicTest().main()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 155, in main
    exit_code = self.shutdown()
                ^^^^^^^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 318, in shutdown
    self.stop_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 583, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 375, in stop_node
    self.stop(wait=wait)
    ^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 205, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
                                  ^^^^^^^^^^^^^^^^^^^^
AssertionError: [node 0] Error: no RPC connection


2/3 - tool_signet_miner.py --descriptors failed, Duration: 60 s

stdout:
2024-04-14T01:47:02.220000Z TestFramework (INFO): PRNG seed is: 1471356017149591392
2024-04-14T01:47:02.221000Z TestFramework (INFO): Initializing test directory /mnt/tmp/test_runner_₿_🏃_20240413_214701/tool_signet_miner_0
2024-04-14T01:48:02.317000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 565, in start_nodes
    node.wait_for_rpc_connection()
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 314, in wait_for_rpc_connection
    self._raise_assertion_error("Unable to connect to bitcoind after {}s".format(self.rpc_timeout))
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 188, in _raise_assertion_error
    raise AssertionError(self._node_msg(msg))
AssertionError: [node 0] Unable to connect to bitcoind after 60s

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 131, in main
    self.setup()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 302, in setup
    self.setup_network()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 396, in setup_network
    self.setup_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 418, in setup_nodes
    self.start_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 568, in start_nodes
    self.stop_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 583, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 375, in stop_node
    self.stop(wait=wait)
    ^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 205, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
                                  ^^^^^^^^^^^^^^^^^^^^
AssertionError: [node 0] Error: no RPC connection
2024-04-14T01:48:02.370000Z TestFramework (INFO): Stopping nodes
[node 0] Cleaning up leftover process


stderr:
Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/tool_signet_miner.py", line 65, in <module>
    SignetMinerTest().main()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 155, in main
    exit_code = self.shutdown()
                ^^^^^^^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 318, in shutdown
    self.stop_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 583, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 375, in stop_node
    self.stop(wait=wait)
    ^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 205, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
                                  ^^^^^^^^^^^^^^^^^^^^
AssertionError: [node 0] Error: no RPC connection


Remaining jobs: [tool_signet_miner.py --legacy-wallet]
3/3 - tool_signet_miner.py --legacy-wallet failed, Duration: 61 s

stdout:
2024-04-14T01:47:02.211000Z TestFramework (INFO): PRNG seed is: 4742527368365497737
2024-04-14T01:47:02.212000Z TestFramework (INFO): Initializing test directory /mnt/tmp/test_runner_₿_🏃_20240413_214701/tool_signet_miner_1
2024-04-14T01:48:02.317000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 565, in start_nodes
    node.wait_for_rpc_connection()
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 314, in wait_for_rpc_connection
    self._raise_assertion_error("Unable to connect to bitcoind after {}s".format(self.rpc_timeout))
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 188, in _raise_assertion_error
    raise AssertionError(self._node_msg(msg))
AssertionError: [node 0] Unable to connect to bitcoind after 60s

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 131, in main
    self.setup()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 302, in setup
    self.setup_network()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 396, in setup_network
    self.setup_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 418, in setup_nodes
    self.start_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 568, in start_nodes
    self.stop_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 583, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 375, in stop_node
    self.stop(wait=wait)
    ^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 205, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
                                  ^^^^^^^^^^^^^^^^^^^^
AssertionError: [node 0] Error: no RPC connection
2024-04-14T01:48:02.370000Z TestFramework (INFO): Stopping nodes
[node 0] Cleaning up leftover process


stderr:
Traceback (most recent call last):
  File "/home/dev/bitcoin/test/functional/tool_signet_miner.py", line 65, in <module>
    SignetMinerTest().main()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 155, in main
    exit_code = self.shutdown()
                ^^^^^^^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 318, in shutdown
    self.stop_nodes()
  File "/home/dev/bitcoin/test/functional/test_framework/test_framework.py", line 583, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 375, in stop_node
    self.stop(wait=wait)
    ^^^^^^^^^
  File "/home/dev/bitcoin/test/functional/test_framework/test_node.py", line 205, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
                                  ^^^^^^^^^^^^^^^^^^^^
AssertionError: [node 0] Error: no RPC connection



TEST                                 | STATUS    | DURATION

feature_signet.py                    | ✖ Failed  | 60 s
tool_signet_miner.py --descriptors   | ✖ Failed  | 60 s
tool_signet_miner.py --legacy-wallet | ✖ Failed  | 61 s

ALL                                  | ✖ Failed  | 181 s (accumulated) 
Runtime: 61 s

dev@bdev01:~/bitcoin$ 

```

kristapsk commented:
Concept ACK

alfonsoromanz commented:
Concept ACK

RandyMcMillan commented:
i concur with @tdb3 - use underscores for signet paths.

RandyMcMillan commented:
I concur with @Sjors.

RandyMcMillan commented:
Ensure the signet path prefix is shared with the wallet so it can find the appropriate wallets for each signet. Treating the global signet as “just another signet” ought to simplify the wallet logic as well as avoiding a bug where the wallet doesn’t distinguish between global signet and custom signets.

RandyMcMillan commented:
You may be able to allow cascading signet
configurations by having a user configured “prefix” setting.
This may also translate better to a gui dialogue later on as well (hot swapping signets from the gui).


[signet]
prune=1000

prefix=<short_string1>
signetchallenge=1234abc…
signetseednode=...

prefix=<short_string2>
signetchallenge=5678def…
signetseednode=...

edilmedeiros reviewed (commented):
Concept ACK.

Code looks good, but not tested yet. 

Any way to add a functional test for this?

BrandonOdiwuor commented:
@Sjors and @RandyMcMillan I love the idea of the `signet_XXXX` config section and would like to work on it if there is a lot of interest in a follow-up PR

BrandonOdiwuor committed:
test: Add test for diffrent signet data directories

- Added a test to verify the Signet data directory when -signetchallenge=51 is correctly set to 'signet_da1745e9'.
- Added a test to verify the main Signet data directory is correctly set to 'signet'.

carnhofdaki commented:
Concept ACK

pablomartin4btc reviewed (commented):
Concept ACK and light tACK 50f4ca2363e008aacb9f8a8aa1bb7abba04c0b33

Tried different challenges and custom `-datadir` option.

tdb3 commented:
This is a great feature.  Looks like the CI logs might be gone?  Maybe this CI could be nudged to run again.  Not sure if this CI failure is related to the one seen a few months ago related to macOS.

luisschwab commented:
Concept ACK. 

Great feature, not being able to do this is a PITA.

carnhofdaki commented:
Concept NACK.

I use a simple script setting datadir according to the current working directory. If the change would be backward-compatible I don't care. But running multiple custom signets and never needed to make a mess. The proposed naming convention (`signet_XXXXXXX`) seems like a mess to me.

carnhofdaki commented:
@edilmedeiros Yes, my harsh word is not being helpful. Just ignore it.

But regarding one network like mainnet there is a signet which anyone can find at mempool.space/signet or just simply running `bitcoind -signet`. Yes it is really something like Signet4 (similar to Testnets) but there is one Signet which was started by original author of the sources and is maintained (blocks are mined+signed) by another Bitcoin Core developer.

Some more related text (by original author mostly) is at https://en.bitcoin.it/wiki/Signet

i-am-yuvi commented:
Concept ACK 

It's a pretty required feature!! I will test it soon!!

i-am-yuvi commented:
Tested ACK f1b40b4de460173e927ebe93cdb5f4ad2ac02859

I tried testing using `./src/bitcoind -signetchallenge=512102ee856c56a5aaadd1656f849bafa4c9dacc86a2878fe546c6189185f842ae2c1851ae`

yuvicc reviewed (approved):
Tested ACK f1b40b4de460173e927ebe93cdb5f4ad2ac02859

- I tried testing using `./src/bitcoind -signetchallenge=512102ee856c56a5aaadd1656f849bafa4c9dacc86a2878fe546c6189185f842ae2c1851ae`
- All the test(functional) passes!!

zaidmstrr reviewed (commented):
Concept ACK
A very much-needed feature when you're working with different signet chains.

jsarenik commented:
How will the default signet directory name be here with this patch?

Could it possibly stay `signet` for the default [Signet](https://en.bitcoin.it/wiki/Signet)? I think the advantages of one default Signet are huge, one of them being the fact you can use [mempool.space/signet](https://mempool.space/signet). There is now a working faucet at [alt.signetfaucet.com](https://alt.signetfaucet.com/).

i-am-yuvi commented:
@jsarenik read the PR description https://github.com/bitcoin/bitcoin/pull/29838#issue-2233837348

jsarenik commented:
> @jsarenik read the PR description https://github.com/bitcoin/bitcoin/pull/29838#issue-2233837348

Yes, it does not seem to honor the main signet. NACK

maflcko reviewed (commented):
are you still working on this?

jsarenik commented on test/functional/tool_signet_miner.py:
```diff
@@ -48,10 +48,11 @@ def run_test(self):
         # generate block with signet miner tool
         base_dir = self.config["environment"]["SRCDIR"]
         signet_miner_path = os.path.join(base_dir, "contrib", "signet", "miner")
+        # breakpoint()
```
Yes, it definitely looks like a stray leftover debug statement. No, I did not write a single line of Python code here. Maybe @ajtowns or @kallewoof may comment on that line.

jsarenik replied:
Apologies for a stray comment. I was out of context "thanks" to GitHub app on Android.

BrandonOdiwuor replied:
@maflcko fixed

BrandonOdiwuor commented:
rebased and removed the leftover debug statement

jsarenik commented:
> @jsarenik @carnhofdaki the behaviour default signet is not changed. I have updated the PR description to include this in case it was unclear.
> 
> This patch only affects custom signets with `signetchallenge` and the goal was to run those on custom directories `signet_XXXXXXXX` so as not to interfere with the default signet which runs on `signet` directory if you run multiple signets

Thank you for this clarification.

utACK abe8639

carnhofdaki commented:
utACK abe8539

BrandonOdiwuor commented:
@maflcko The assumption was, providing default signet challenge is not so useful as it would result in the loss of default the signet configurations, hence did not include the check

https://github.com/bitcoin/bitcoin/blob/2db00278ea571d0f734609f9fefecfa75c16ee34/src/kernel/chainparams.cpp#L393-L408

BrandonOdiwuor commented:
> Also, the CI fails currently.

rebased and fixed the ci error

flack commented on src/common/args.h:
```diff
@@ -212,6 +212,14 @@ class ArgsManager
 
+    /**
+     * Get signet data directory path.
+     * If a signet-challange argument is provided, it is used in constructing the directory path.
```
nit: challange => challenge

jsarenik commented:
utACK 6913904

RandyMcMillan commented:
👀 going to revisit this PR today...

yuvicc reviewed (commented):
re-ACK 69139049194e55a7fd99b7125b9afccb8585922d

achow101 reviewed (commented):
Concept ACK

BrandonOdiwuor commented on test/functional/feature_signet.py:
```diff
@@ -101,6 +102,23 @@ def check_getmininginfo(node_idx, signet_idx):
+            self.log.info(f"Checking node datadir: {datadir}")
+            assert_equal(path.basename(datadir), expected_dirname)
+
+            assert datadir.is_dir() # check if the direcory exists
```
fixed

RandyMcMillan commented:
note: i already tested 

https://github.com/bitcoin-core/gui/pull/896/files

on top of this PR 🔥

achow101 commented:
Are you still working on this? This has needed rebase for a few months now.

ekzyis reviewed (commented):
tACK bf2d4bf988 I ran all the functional tests

As with `rpc_url` and `get_auth_cookie`, I think `delete_cookie_file` in `test_node.py:287` also needs to be passed `self.get_chain()` now. I tried to find other cases by searching for variations of `datadir_path` but didn't find any.

I also agree with @RandyMcMillan: I don't think we need `hash160` here.

I left a separate comment about `self.get_chain()`. I think the naming is confusing given that `self.chain` also exists.

@BrandonOdiwuor, are you still planning to work on this? If not, I’d be happy to continue your work with proper credit in a new PR based on this one, unless there's a different preferred etiquette.
